File: src/main/java/ca/rafemurray/StartBlockEntity.java
package ca.rafemurray;

import net.minecraft.block.BlockState;
import net.minecraft.block.entity.BlockEntity;
import net.minecraft.entity.Entity;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.nbt.NbtCompound;
import net.minecraft.registry.Registry;
import net.minecraft.registry.RegistryWrapper;
import net.minecraft.util.Identifier;
import net.minecraft.util.math.BlockPos;

import java.util.UUID;

public class StartBlockEntity extends BlockEntity {
    private UUID courseId;
    private Course course;

    public StartBlockEntity(BlockPos pos, BlockState state) {
        super(ModBlockEntities.START_BLOCK_ENTITY, pos, state);
        this.courseId = null;
        this.course = null;
    }

    public void setCourse(Course course) {
        this.courseId = course.getId();
        this.course = course;
        markDirty();
    }
    public void onSteppedOn(Entity entity) {
        if (entity instanceof PlayerEntity && course != null && course.isValid()) {
            ElytraRacingPlayerEntity player = (ElytraRacingPlayerEntity) entity;
            player.startCourse(course);
        }
    }

//    @Override
//    protected void readNbt(NbtCompound tag, RegistryWrapper.WrapperLookup registries) {
//
//    }
}

File: src/main/java/ca/rafemurray/block/entity/StartBlockEntity.kt
package ca.rafemurray.block.entity

import ca.rafemurray.ElytraRacingPlayer
import ca.rafemurray.course.Course
import ca.rafemurray.course.Courses.Companion.instance
import net.minecraft.core.BlockPos
import net.minecraft.core.UUIDUtil
import net.minecraft.world.entity.Entity
import net.minecraft.world.level.block.entity.BlockEntity
import net.minecraft.world.level.block.state.BlockState
import net.minecraft.world.level.storage.ValueInput
import net.minecraft.world.level.storage.ValueOutput
import java.util.*

class StartBlockEntity(pos: BlockPos, state: BlockState) :
    BlockEntity(ModBlockEntities.START_BLOCK_ENTITY, pos, state) {
    private var courseId: UUID? = null

    fun setCourse(course: Course) {
        this.courseId = course.id
        setChanged()
    }

    fun stepOn(entity: Entity) {
        if (entity is ElytraRacingPlayer && courseId != null) {
            val course = instance.getCourse(courseId)
            if (course == null || !course.isValid || course === entity.course) {
                return
            }
            entity.startCourse(course)
        }
    }

    override fun loadAdditional(input: ValueInput) {
        val optionalCourseId = input.read(
            Course.COURSE_ID_IDENTIFIER.toString(),
            UUIDUtil.CODEC
        )
        courseId = optionalCourseId.orElse(null)
        super.loadAdditional(input)
    }


    override fun saveAdditional(out: ValueOutput) {
        out.storeNullable(
            Course.COURSE_ID_IDENTIFIER.toString(),
            UUIDUtil.CODEC,
            courseId
        )
        super.saveAdditional(out)
    }
}
